timelock-impl/src/main/java/com/palantir/atlasdb/timelock/ConjureTimelockResource.java

```diff
@@ -187,22 +193,57 @@ private static Set<LockDescriptor> fromConjureLockDescriptors(Set<ConjureLockDes
             AuthHeader authHeader, String namespace, ConjureRefreshLocksRequest request) {
-        return handleExceptions(() -> Futures.transform(
-                forNamespace(namespace).refreshLockLeases(fromConjureLockTokens(request.getTokens())),
+        return refreshLocksInternal(
+                namespace,
+                fromConjureLockTokens(request.getTokens()),
                 refreshed -> ConjureRefreshLocksResponse.of(
-                        toConjureLockTokens(refreshed.refreshedTokens()), refreshed.getLease()),
+                        toConjureLockTokens(refreshed.refreshedTokens()), refreshed.getLease()));
+    }
+
+    @Override
+    public ListenableFuture<ConjureRefreshLocksResponseV2> refreshLocksV2(
+            AuthHeader authHeader, String namespace, ConjureRefreshLocksRequestV2 request) {
+        return refreshLocksInternal(
+                namespace,
+                fromConjureLockV2Tokens(request.getTokens()),
+                refreshed -> ConjureRefreshLocksResponseV2.of(
+                        toConjureLockV2Tokens(refreshed.refreshedTokens()), refreshed.getLease()));
+    }
+
+    private <T> ListenableFuture<T> refreshLocksInternal(
+            String namespace, Set<LockToken> tokens, Function<RefreshLockResponseV2, T> userTokenTranslator) {
+        return handleExceptions(() -> Futures.transform(
+                forNamespace(namespace).refreshLockLeases(tokens),
+                userTokenTranslator::apply,
                 MoreExecutors.directExecutor()));
     }
 
     @Override
     public ListenableFuture<ConjureUnlockResponse> unlock(
             AuthHeader authHeader, String namespace, ConjureUnlockRequest request) {
+        return unlockInternal(
+                namespace,
+                fromConjureLockTokens(request.getTokens()),
+                unlocked -> ConjureUnlockResponse.of(toConjureLockTokens(unlocked)));
+    }
+
+    @Override
+    public ListenableFuture<ConjureUnlockResponseV2> unlockV2(
+            AuthHeader authHeader, String namespace, ConjureUnlockRequestV2 request) {
+        return unlockInternal(
+                namespace,
+                fromConjureLockV2Tokens(request.getTokens()),
+                unlocked -> ConjureUnlockResponseV2.of(toConjureLockV2Tokens(unlocked)));
+    }
+
+    private <T> ListenableFuture<T> unlockInternal(
+            String namespace, Set<LockToken> tokens, Function<Set<LockToken>, T> userTokenTranslator) {
         return handleExceptions(() -> Futures.transform(
-                forNamespace(namespace).unlock(fromConjureLockTokens(request.getTokens())),
-                unlocked -> ConjureUnlockResponse.of(toConjureLockTokens(unlocked)),
-                MoreExecutors.directExecutor()));
+                forNamespace(namespace).unlock(tokens), userTokenTranslator::apply, MoreExecutors.directExecutor()));
     }
 
+    // The reason for duplication in the following four methods is for performance reasons; while one might
```

Comment:
This is fine, though it would be interesting to benchmark it and see if the perf difference is noticeable -- from what I've read, I wouldn't be surprised if it's equally performant

Reply:
Yep. I don't imagine the difference will be large.